src/kit.ts

```diff
@@ -1108,65 +1108,72 @@ export async function descriptionForKit(kit: Kit, shortVsName: boolean = false):
-                generator: '${generator}',
-                userHome: paths.userHome,
-                workspaceFolder: '${workspaceFolder}',
-                workspaceFolderBasename: '${workspaceFolderBasename}',
-                workspaceHash: '${workspaceHash}',
-                workspaceRoot: '${workspaceRoot}',
-                workspaceRootFolderName: '${workspaceRootFolderName}'
-            }
-        });
-    }
-    return kit;
-}
-
-export async function readKitsFile(filepath: string): Promise<Kit[]> {
-    const fileStats = await fs.tryStat(filepath);
+export async function readKitsFile(filePath: string, workspaceFolder?: string, opts?: expand.ExpansionOptions): Promise<Kit[]> {
+    const fileStats = await fs.tryStat(filePath);
     if (!fileStats) {
-        log.debug(localize('not.reading.nonexistent.kit', 'Not reading non-existent kits file: {0}', filepath));
+        log.debug(localize('not.reading.nonexistent.kit', 'Not reading non-existent kits file: {0}', filePath));
         return [];
     }
     if (!fileStats.isFile()) {
-        log.debug(localize('not.reading.invalid.path', 'Not reading invalid kits file: {0}', filepath));
+        log.debug(localize('not.reading.invalid.path', 'Not reading invalid kits file: {0}', filePath));
         return [];
     }
-    log.debug(localize('reading.kits.file', 'Reading kits file {0}', filepath));
-    const content_str = await fs.readFile(filepath);
+    log.debug(localize('reading.kits.file', 'Reading kits file {0}', filePath));
+    const content_str = await fs.readFile(filePath);
     let kits_raw: object[] = [];
     try {
         kits_raw = json5.parse(content_str.toLocaleString());
     } catch (e) {
-        log.error(localize('failed.to.parse', 'Failed to parse {0}: {1}', path.basename(filepath), util.errorToString(e)));
+        log.error(localize('failed.to.parse', 'Failed to parse {0}: {1}', path.basename(filePath), util.errorToString(e)));
         return [];
     }
     const validator = await loadSchema('schemas/kits-schema.json');
     const is_valid = validator(kits_raw);
     if (!is_valid) {
         const errors = validator.errors!;
-        log.error(localize('invalid.file.error', 'Invalid kit contents {0} ({1}):', path.basename(filepath), filepath));
+        log.error(localize('invalid.file.error', 'Invalid kit contents {0} ({1}):', path.basename(filePath), filePath));
         for (const err of errors) {
             log.error(` >> ${err.dataPath}: ${err.message}`);
         }
         return [];
     }
     const kits = kits_raw as Kit[];
-    log.info(localize('successfully.loaded.kits', 'Successfully loaded {0} kits from {1}', kits.length, filepath));
-    return Promise.all(dropNulls(kits).map(expandKitVariables));
+    log.info(localize('successfully.loaded.kits', 'Successfully loaded {0} kits from {1}', kits.length, filePath));
+
+    const expandedKits: Kit[] = [];
+    for (const kit of kits) {
```

Comment:
The previous implementation removed null kits before expanding variables.  I don't know the case where a kit would be null, but I think we should preserve the behavior just in case.